x-pack/legacy/plugins/lens/public/legacy.ts

```diff
@@ -14,6 +14,7 @@ import { plugin } from './index';
 const pluginInstance = plugin();
 pluginInstance.setup(npSetup.core, {
   ...npSetup.plugins,
+  uiActions: npStart.plugins.uiActions,
```

Comment:
This does not look correct. We cannot (technically in legacy platform we can, but it would break as soon as we move to Kibana platform) inject a start dependency into the setup phase of the plugin (that runs before the start phase).

If we get access to the `uiActions` only via the start contract, we need to pass it down from the start method of our plugin instead. We can use the `createGetterSetter` utils in `kibana_utils` to store a reference of it in start and use it later when clicking.

Reply:
yup, thats right ... general problem, what to do when you need to register something to a registry (needs to happen in setup phase) but the thing you are registering has a start dependencies.

what we usually do is
- create `services.ts` file and create uiActions getter and setter (check visualizations plugin, data plugin, expression plugin, ...)
- wherever you need the uiActions use the getter
- in your plugins start lifecycle method use the setter to set the reference 
- make sure you don't expose anything using the getter as a static export

Reply:
We have the same problem with “formatFactory” (currently importer via legacy imports, but actually exposed in the start phase. In a separate Pr I solved this problem by passing down a promode of the contract and resolve if needed

Reply:
Yea, I've been solving this @flash1293's way, which I think is the correct way, it follows the pattern of `coreSetup.getStartServices`. So what would get passed down to handlers I think is `async getUiActionsService()` and that can be something like:

 ```
const getUiActionsService: async () =>  await coreSetup.getStartServices()[1].uiActions;
```

Expression fns are async so I think this should work fine.  I think we should prefer that pattern over the getter/setter. 

cc @joshdover